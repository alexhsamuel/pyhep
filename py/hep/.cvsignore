*.pyc
*.pyo
config.py
*-install
pdt.pickle
