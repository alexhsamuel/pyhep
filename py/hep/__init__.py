#-----------------------------------------------------------------------
#
# module hep
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Module for general-purpose code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

import os
import signal
import traceback

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TimeoutException(Exception):
    """A timeout occurred."""

    pass



class NotSupportedError(RuntimeError):
    """A feature is not supported."""

    pass



class SyntaxError(Exception):

    pass



#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Token:

    pass



class Timeout:

    def __timeout_handler(self, signal_number, frame):
        assert signal_number == signal.SIGALRM
        signal.signal(signal.SIGALRM, self.__old_handler)
        self.__active = 0
        raise TimeoutException, self.__message


    def __init__(self, time, message):
        self.__message = message
        self.__old_handler = \
            signal.signal(signal.SIGALRM, self.__timeout_handler)
        self.__active = 1
        signal.alarm(time)


    def cancel(self):
        if not self.__active:
            return
        signal.alarm(0)
        signal.signal(signal.SIGALRM, signal.SIG_DFL)
        self.__active = 0



class MappingAsObject:
    """Implement the object attribute protocol using a mapping.

    A 'MappingAsObject' wraps an object satifying the mapping protocol.
    Attribute references are converted to item gets and sets in the
    underlying mapping."""

    def __init__(self, mapping):
        """Wrap 'mapping' in the object procotol."""
        
        self.__dict__["_MappingAsObject__mapping"] = mapping


    _mapping = property(lambda self: self.__mapping)
    """Read-only access to the underlying mapping."""


    def __getattr__(self, name):
        try:
            return self.__mapping[name]
        except KeyError:
            raise AttributeError, name


    def __setattr__(self, name, value):
        try:
            self.__mapping[name] = value
        except KeyError, exception:
            raise AttributeError, str(exception)



class NrBreitWigner:
    """A non-resonant Breit-Wigner line shape amplitude factor."""

    def __init__(self, mass, width):
        """Construct a non-relativistic Breit-Wigner function.

        'mass' -- The resonance mass.
        
        'width' -- The resonance width."""

        self.__mass = mass
        self.__width = width


    def __call__(self, s):
        """Return the amplitude at energy-squared 's'."""

        mass = self.__mass
        return 1 / (s - complex(mass * mass, mass * self.__width))



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

try:
    enumerate = __builtins__.enumerate
except AttributeError:
    def enumerate(iterable):
        count = 0
        for item in iter(iterable):
            yield count, item
            count += 1


try:
    sorted = __builtins__.sorted
except AttributeError:
    def sorted(iterable, cmp=cmp):
        result = list(iterable)
        result.sort(cmp)
        return result


try:
    reversed = __builtins__.reversed
except AttributeError:
    def reversed(iterable):
        result = list(iterable)
        result.reverse()
        return result
    

def sort(sequence, sort_fn=None):
    """Return a copy of 'sequence', sorted by 'sort_fn'.

    'sort_fn' -- The comparison function to use.  If 'None', the default
    (built-in 'cmp') is used."""

    sequence = list(sequence)
    if sort_fn is None:
        sort_fn = cmp
    sequence.sort(sort_fn)
    return sequence


def binarySearch(sequence, value, compare=cmp):
    """Find 'value' in sorted 'sequence' by binary search.

    'sequence' -- A sequence of values.  It is assumed to be sorted by
    'compare'.

    'compare' -- A comparison function.

    returns -- The index of the largest element in 'sequence' smaller
    than 'value', or -1 if there is no such element."""

    if len(sequence) == 0:
        return -1
    if compare(value, sequence[0]) < 0:
        return -1
    if compare(value, sequence[-1]) > 0:
        return len(squence) - 1

    i0 = 0
    i2 = len(sequence) - 1
    while True:
        if i2 <= i0 + 1:
            return i2 - 1
        i1 = (i0 + i2) // 2
        if compare(value, sequence[i1]) < 0:
            i2 = i1
        else:
            i0 = i1


def mapFromKeys(key_sequence, value_function):
    """Construct a map from a sequence of keys.

    A map is created whose keys are the elements of 'key_sequence'.  For
    each key, the corresponding value is constructed from the key using
    'value_function'.  If 'sequence' contains a particular key more than
    once, only one is placed in the map (which one is undefined).

    'key_sequence' -- A sequence of key values.

    'value_function' -- A callable for determining the value for each
    element of the sequence.  It is invoked for each key, passing the
    kay as the single argument."""
    
    result = {}
    for key in key_sequence:
        result[key] = value_function(key)
    return result


def mapFromValues(key_function, value_sequence):
    """Construct a map from a sequence of values.

    A map is created whose values are the elements of 'value_sequence'.
    For each value, the corresponding key is constructed from the value
    using 'key_function'.  If 'key_function' assigns the same key to
    more than one element of the sequence, only one element is actually
    used in the map (which one is undefined).

    'key_function' -- A callable for determining the key for each
    element of the sequence.  It is invoked for each value, passing
    the value as the single argument.

    'value_sequence' -- A sequence of values."""
    
    result = {}
    for value in value_sequence:
        result[key_function(value)] = value
    return result


def invertMap(map):
    """Produce a dictionary by inverting keys and values of 'map'."""

    return dict([ (value, key) for key, value in map.items() ])


def remove(dict, key):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    returns -- The value corresponding to 'key' in 'dict'.  

    raises -- 'KeyError' if 'dict' has no key 'key'."""
    
    value = dict[key]
    del dict[key]
    return value


def popkey(map, key, default=None):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    'default' -- The value to return if 'key' is not in 'map'.

    returns -- The value corresponding to 'key' in 'dict', or 'default'
    if none."""

    if key in map:
        result = map[key]
        del map[key]
        return result
    else:
        return default


def getHostName():
    """Return the name of this host."""

    import socket

    try:
        return socket.gethostbyname_ex(socket.gethostname())[0]
    except socket.error:
        return "(unknown)"


def isMap(object):
    """Determine whether an object satisfies the mapping protocol.

    returns -- True if 'object' satisfies the map protocol as
    specified.""" 

    # FIXME.  There probably is a more consistent way to do this.

    # If the object is an instance of the standard 'dict' class (or a
    # subclass of it), it must satisfy the mapping protocol.
    if isinstance(object, dict):
        return True
    # Check for the required methods.
    for method_name in ("get", "__getitem__", "__len__", ):
        if not hasattr(object, method_name):
            return False
    # The object checks out.
    return 1


def mkdirRecursively(path):
    """Create a directory, including parent directories if necessary."""

    parent, name = os.path.split(path)
    if parent != "" and not os.path.isdir(parent):
        mkdirRecursively(parent)
    os.mkdir(path)


def rmRecursively(path):
    """Remove the file or directory at 'path'.

    If 'path' is a directory, its contents are removed first."""

    if os.path.isdir(path):
        # Construct a list of contents of the directory.
        contents = map(lambda e, path=path: os.path.join(path, e),
                       os.listdir(path))
        # Remove them all.
        map(rmRecursively, contents)
        # Remove the directory itself.
        os.rmdir(path)
    else:
        os.unlink(path)


