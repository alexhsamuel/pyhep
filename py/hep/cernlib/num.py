#-----------------------------------------------------------------------
#
# num.py
#
# Copyright (C) 2004 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Numerical functions based on CERNLIB."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import ext

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def chiSquareCDF(x, n):
    """CDF for chi-square distribution.

    'n' -- The chi-squared parameter; must be one or larger.

    returns -- The probability that a random variable chosen
    according to a chi-square distribution with 'n' degrees of
    freedom will have a value of 'x' or lower."""

    return 1 - ext.prob(x, n)


