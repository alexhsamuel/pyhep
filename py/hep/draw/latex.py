#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

from   __init__ import *
from   fonts import *

#-----------------------------------------------------------------------
# constants
#-----------------------------------------------------------------------

math_rm_font = "Times"
math_it_font = "Times-Italic"


# Text-mode escapes that are printed literally.
literal_escapes = [ "\\#", "\\%", "\\\\", "\\{", "\\}", "\\&", "\\$", ]


# Math-mode spacing escapes mapped to a faction of the standard space
# width. 
math_spacing_escapes = {
    "\\,": 0.4,
    "\\:": 0.6,
    "\\;": 0.8,
    "\\ ": 1.0,
    }


# Math-mode escapes to be printed literally in roman text.
math_roman_escapes = (
    "\\arccos",
    "\\arcsin",
    "\\arctan",
    "\\arg",
    "\\cos",
    "\\cosh",
    "\\cot",
    "\\coth",
    "\\csc",
    "\\deg",
    "\\det",
    "\\dim",
    "\\exp",
    "\\gcd",
    "\\hom",
    "\\inf",
    "\\ker",
    "\\lg",
    "\\lim",
    "\\liminf",
    "\\limsup",
    "\\ln",
    "\\log",
    "\\max",
    "\\min",
    "\\Pr",
    "\\sec",
    "\\sin",
    "\\sinh",
    "\\sup"
    "\\tan",
    "\\tanh",
    )


# Characters that should be displayed as symbols in math mode.  
math_symbols = {
    "-":              "minus",
    }


# Characters that should be displayed in italics in math mode.
math_italic_chars = "abcdefghijklmnopqrstuvwxyz" \
                    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"


# Map math-mode escapes to symbol names and whether to display them in
# italics. 
math_symbol_escapes = {
    "\\alpha":        ("alpha", True),
    "\\beta":         ("beta", True),
    "\\gamma":        ("gamma", True),
    "\\delta":        ("delta", True),
    "\\epsilon":      ("epsilon", True),
    "\\zeta":         ("zeta", True),
    "\\eta":          ("eta", True),
    "\\theta":        ("theta", True),
    "\\vartheta":     ("vartheta", True),
    "\\iota":         ("iota", True),
    "\\kappa":        ("kappa", True),
    "\\lambda":       ("lambda", True),
    "\\mu":           ("mu", True),
    "\\nu":           ("nu", True),
    "\\xi":           ("xi", True),
    "\\omicron":      ("omicron", True),
    "\\pi":           ("pi", True),
    "\\varpi":        ("varpi", True),
    "\\rho":          ("rho", True),
    "\\varsigma":     ("varsigma", True),
    "\\sigma":        ("sigma", True),
    "\\tau":          ("tau", True),
    "\\upsilon":      ("upsilon", True),
    "\\phi":          ("phi", True),
    "\\varphi":       ("varphi", True),
    "\\chi":          ("chi", True),
    "\\psi":          ("psi", True),
    "\\omega":        ("omega", True),

    "\\Alpha":        ("Alpha", True),
    "\\Beta":         ("Beta", True),
    "\\Gamma":        ("Gamma", True),
    "\\Delta":        ("Delta", True),
    "\\Epsilon":      ("Epsilon", True),
    "\\Zeta":         ("Zeta", True),
    "\\Eta":          ("Eta", True),
    "\\Theta":        ("Theta", True),
    "\\Iota":         ("Iota", True),
    "\\Kappa":        ("Kappa", True),
    "\\Lambda":       ("Lambda", True),
    "\\Mu":           ("Mu", True),
    "\\Nu":           ("Nu", True),
    "\\Xi":           ("Xi", True),
    "\\Omicron":      ("Omicron", True),
    "\\Pi":           ("Pi", True),
    "\\Rho":          ("Rho", True),
    "\\Sigma":        ("Sigma", True),
    "\\Tau":          ("Tau", True),
    "\\Upsilon":      ("Varupsilon", True),
    "\\Phi":          ("Phi", True),
    "\\Chi":          ("Chi", True),
    "\\Psi":          ("Psi", True),
    "\\Omega":        ("Omega", True),

    "\\Im":           ("Im", False),
    "\\Re":           ("Re", False),
    "\\approx":       ("approx", False),
    "\\ast":          ("ast", False),
    "\\bullet":       ("bullet", False),
    "\\cdot":         ("cdot", False),
    "\\dagger":       ("dagger", False),
    "\\ddagger":      ("ddagger", False),
    "\\degree":       ("degree", False),
    "\\del":          ("del", False),
    "\\div":          ("div", False),
    "\\downarrow":    ("downarrow", False),
    "\\equiv":        ("equiv", False),
    "\\geq":          ("geq", False),
    "\\infty":        ("infty", False),
    "\\leftarrow":    ("leftarrow", False),
    "\\leq":          ("leq", False),
    "\\nabla":        ("nabla", False),
    "\\neq":          ("neq", False),
    "\\pm":           ("pm", False),
    "\\prime":        ("prime", False),
    "\\propto":       ("propto", False),
    "\\rightarrow":   ("rightarrow", False),
    "\\sim":          ("sim", False),
    "\\times":        ("times", False),
    "\\uparrow":      ("uparrow", False),
}


#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class PeekIterator(object):

    def __init__(self, iterator):
        self.__iterator = iterator
        self.peek = None
        self.is_finished = False
        self.__getNext()


    def __getNext(self):
        try:
            self.peek = self.__iterator.next()
        except StopIteration:
            self.is_finished = True
            self.peek = None


    def next(self):
        if self.is_finished:
            raise StopIteration
        else:
            element = self.peek
            self.__getNext()
            return element
        


#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def isLetter(character):
    code = ord(character)
    return 65 <= code <= 90 or 97 <= code <= 122 


def tokenize(text):
    i = 0
    length = len(text)

    while i < length:
        if text[i] == "\\":
            token = text[i : i + 2]
            i += 2
            if isLetter(token[1]):
                while i < length and isLetter(text[i]):
                    token += text[i]
                    i += 1
            yield token
        elif text[i] == "%":
            i += 1
            while i < length and text[i] != "\n":
                i += 1
            i += 1
            continue
        elif text[i] in " \t\n":
            linefeeds = 0
            while i < length and text[i] in " \t\n":
                if text[i] == "\n":
                    linefeeds += 1
                i += 1
            if linefeeds > 1:
                yield "\n"
            else:
                yield " "
        elif isLetter(text[i]):
            token = ""
            while i < length and isLetter(text[i]):
                token += text[i]
                i += 1
            yield token
        else:
            token = text[i]
            i += 1
            yield token


def renderMath(renderer, x, y, tokens, font_size, style):
    # Handle blocks.
    if tokens.peek == "{":
        tokens.next()
        x, y = renderMathBlock(
            renderer, x, y, tokens, font_size, style)
        if tokens.next() != "}":
            raise ParseError, "missing closing }"

    # Handle superscripts.
    elif tokens.peek == "^":
        tokens.next()
        # Apply poor-man's italic correction.
        # FIXME: Do this better.
        x += 0.1 * font_size
        x, ignore_y = renderMath(
            renderer, x, y + 0.35 * font_size, tokens, 0.8 * font_size,
            style)

    # Handle subscripts.
    elif tokens.peek == "_":
        tokens.next()
        x, ignore_y = renderMath(
            renderer, x, y - 0.25 * font_size, tokens, 0.8 * font_size,
            style)

    # Handle characters that are drawn as symbols.
    elif tokens.peek in math_symbols:
        token = tokens.next()
        symbol_name = math_symbols[token]
        renderer.font = (math_rm_font, font_size)
        renderer.symbol((x, y), symbol_name)
        x += getSymbolExtent(symbol_name, renderer.font)[0]

    # Handle escapes.
    elif tokens.peek[0] == "\\":
        escape = tokens.next()

        if escape in math_spacing_escapes:
            width = math_spacing_escapes[escape]
            font = (math_rm_font, font_size)
            x += width * getTextExtent(" ", font)[0]

        # Handle escapes that are drawn as roman text.
        elif escape in math_roman_escapes:
            text = escape[1 :] + " "
            renderer.font = (math_rm_font, font_size)
            renderer.text((x, y), text, (0, 0))
            x += getTextExtent(text, renderer.font)[0]

        # Handle escapes that are drawn as symbols.
        elif escape in math_symbol_escapes:
            symbol_name, italic = math_symbol_escapes[escape]
            if italic:
                font_family = math_it_font
            else:
                font_family = math_rm_font
            renderer.font = (font_family, font_size)
            renderer.symbol((x, y), symbol_name)
            x += getSymbolExtent(symbol_name, renderer.font)[0]

        elif escape in literal_escapes:
            text = escape[1 :]
            renderer.font = (math_rm_font, font_size)
            renderer.text((x, y), text, (0, 0))
            x += getTextExtent(text, renderer.font)[0]

        else:
            raise ParseError, "unrecognized escape '%s'" % escape

    # Skip whitespace.
    elif tokens.peek in (" ", "\n", ):
        tokens.next()
        
    # Write out everything else.
    else:
        text = tokens.next()
        if isLetter(text[0]):
            # Alphabetic text is written in italics.
            font_family = math_it_font
        else:
            # Everything else in roman.
            font_family = math_rm_font
        renderer.font = (font_family, font_size)
        renderer.text((x, y), text, (0, 0))
        x += getTextExtent(text, renderer.font)[0]

    return x, y
            

def renderMathBlock(renderer, x, y, tokens, font_size, style):
    while True:
        if tokens.is_finished:
            raise ParseError, "missing closing $"
        elif tokens.peek in ("$", "}", ):
            return x, y
        else:
            x, y = renderMath(renderer, x, y, tokens, font_size, style)
    return x, y


def renderDash(renderer, x, y, tokens, font, style):
    assert tokens.peek == "-"
    tokens.next()
    if tokens.peek == "-":
        tokens.next()
        if tokens.peek == "-":
            tokens.next()
            symbol = "emdash"
        else:
            symbol = "endash"
    else:
        symbol = "hyphen"
    renderer.font = font
    renderer.symbol((x, y), symbol)
    font_family, font_size = font
    x += getSymbolExtent(symbol, font)[0]
    return x, y


def renderBraceBlock(renderer, x, y, tokens, font, style):
    token = tokens.next()
    assert token == "{"
    x, y = renderBlock(renderer, x, y, tokens, font, style)
    token = tokens.next()
    if token != "}":
        raise ParseError, "missing closing }"
    return x, y


def renderBlock(renderer, x, y, tokens, font, style):
    font_family, font_size = font

    while True:

        if tokens.is_finished or tokens.peek == "}":
            return x, y

        elif tokens.peek == "{":
            x, y, = renderBraceBlock(renderer, x, y, tokens, font, style)

        elif tokens.peek == "$":
            tokens.next()
            x, y = renderMathBlock(
                renderer, x, y, tokens, font_size, style)
            if tokens.next() != "$":
                raise ParseError, "missing closing $"

        elif tokens.peek == "-":
            x, y = renderDash(
                renderer, x, y, tokens, font, style)

        elif tokens.peek[0] == "\\":
            token = tokens.next()

            if token in literal_escapes:
                text = token[1 :]
                renderer.font = font
                renderer.text((x, y), text, (0, 0))
                x += getTextExtent(text, font)[0]
                
        else:
            token = tokens.next()
            renderer.font = font
            renderer.text((x, y), token, (0, 0))
            x += getTextExtent(token, font)[0]


class ExtentsRenderer:

    def __init__(self):
        self.__x_min = None
        self.__y_min = None
        self.__x_max = None
        self.__y_max = None
        self.transformation_stack = []
        self.__font = None


    def __include(self, x, y):
        if self.__x_min is None or x < self.__x_min:
            self.__x_min = x
        if self.__x_max is None or x > self.__x_max:
            self.__x_max = x
        if self.__y_min is None or y < self.__y_min:
            self.__y_min = y
        if self.__y_max is None or y > self.__y_max:
            self.__y_max = y


    def transform(self, transformation):
        self.transformation_stack.insert(0, transformation)


    def untransform(self):
        self.transformation_stack.pop(0)


    def map(self, point):
        for transformation in self.transformation_stack:
            point = transformation(point)
        return point


    def text(self, position, text, alignment):
        x, y = self.map(position)
        x_alignment, y_alignment = alignment
        width, height, depth = getTextExtent(text, self.font)
        self.__include(x - x_alignment * width,
                       y - y_alignment * height)
        self.__include(x + (1 - x_alignment) * width,
                       y + (1 - y_alignment) * height)


    def symbol(self, position, symbol):
        x, y = self.map(position)
        width, height, depth = getSymbolExtent(symbol, self.font)
        self.__include(x, y)
        self.__include(x + width, y + height)


    extents = property(lambda self: (self.__x_min, self.__y_min,
                                     self.__x_max, self.__y_max))



def render(renderer, position, text, style, alignment):
    if text == "":
        return
    
    font_family = math_rm_font
    font_size = style.get("font_size", 10 * point)
    font = (font_family, font_size)
    color = style.get("color", colors["black"])

    x, y = position
    if alignment != (0, 0):
        # Compute the extents of the text by rendering it.
        tokens = PeekIterator(tokenize(text))
        extents_renderer = ExtentsRenderer()
        renderBlock(extents_renderer, x, y, tokens, font, style)
        x0, y0, x1, y1 = extents_renderer.extents
        
        # Align text by adjusting its position.
        x_alignment, y_alignment = alignment
        # For x alignment, use the rendered width of the text.
        x -= x_alignment * (x1 - x0)
        # For y alignment, estimate the character height of the font
        # size.
        y -= y_alignment * 0.8 * font_size

    renderer.color = color
    tokens = PeekIterator(tokenize(text))
    renderBlock(renderer, x, y, tokens, font, style)

    if tokens.peek == "}":
        raise ParseError, "extra closing }"


def scientificNotation(value, decimal_places):
    if value == 0:
        return "0"
    scale = log10(abs(value))
    exponent = int(floor(scale))
    mantissa = value / 10 ** exponent
    return r"%.*f\times 10^{%d}" % (decimal_places, mantissa, exponent)


def formatStatistic(statistic, value_digits=4, error_digits=None):
    """Return a nice string representation of 'statistic'.

    returns -- LaTeX markup for 'statistic'.  Must be used in math mode."""

    if error_digits is None:
        error_digits = int(ceil(value_digits / 2)) + 1
    return r"%s\pm%s" \
           % (formatNumber(statistic, value_digits),
              formatNumber(statistic.uncertainty, error_digits))


def formatNumber(number, significant_digits=4):
    """Return a nice string representation of floating-point 'number'."""

    number = float(number)

    # If all the significant digits are to the left of the decimal
    # point (or if 'number' is an integer), use an integer
    # representation. 
    if abs(number) > (10 ** (significant_digits - 1)) \
       or int(number) == number:
        return "%d" % int(round(number))

    if abs(number) < 1e-4:
        # Format small numbers in exponential notation.
        result = scientificNotation(number, significant_digits)
    else:
        # For other numbers, figure out the required precision.
        scale = int(log10(abs(number)))
        if abs(number) < 1:
            decimal_places = significant_digits - scale
        else:
            decimal_places = significant_digits - scale - 1
        result = ("%%.%df" % decimal_places) % number
    # Don't end with a decimal point.
    if result[-1] == '.':
        result = result[:-1]

    return result


