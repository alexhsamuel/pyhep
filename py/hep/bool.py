#-----------------------------------------------------------------------
#
# module hep.bool
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Implementation of bool for older Python versions.

This module provides 'bool', 'True', and 'False' similar to those found
in recent versions of Python.  These symbols are defined only when
Python doesn't include its own built-in."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.ext import BoolArray

#-----------------------------------------------------------------------

if not hasattr(__builtins__, "bool"):

    False = 0

    True = 1


    class bool(object):

        def __new__(class_, value):
            if value:
                return True
            else:
                return False



