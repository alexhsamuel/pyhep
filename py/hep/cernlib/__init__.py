#-----------------------------------------------------------------------
#
# module hep.cernlib
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""CERNLIB library interface."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import ext
from   hep import popkey
from   hep.bool import *

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def integrate(function, *region, **kw_args):
    """Numerically integrate 'function'.

    Performs multidimensional numerical integration of 'function' over a
    rectangular integration 'region'.

    'function' -- A callable or expression.

    '*region' -- One-dimensional ranges specifying a rectangular
    integration region.  Each range is a tuple of the form '(var_name,
    lo, hi)', where 'var_name' is the name of the integration variable
    and 'lo' and 'hi' are 'float' values specifying the interval to
    integrate over.

    'accuracy' -- Specify the desired accuracy with this keyword
    argument.

    returns -- The integral value."""

    # If 'function' isn't already callable, treat it as an expression. 
    import hep.expr
    if not callable(function):
        function = hep.expr.asExpression(function)

    if hep.expr.isExpression(function) \
       and not hep.expr.isCompiledExpression(function):
        function = hep.expr.setTypesFixed(function, None, float)
        function = hep.expr.compile(function)

    accuracy = float(popkey(kw_args, "accuracy", 1e-8))
    if len(kw_args) > 0:
        raise TypeError, \
              "'integrate' got an unexpected keyword argument '%s'" \
              % kw_args.keys()[0]

    # Perform the integration.
    return ext.integrate(function, region, accuracy)


