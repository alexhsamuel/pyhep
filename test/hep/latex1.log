This is pdfeTeX, Version 3.141592-1.21a-2.2 (Web2C 7.5.4) (format=latex 2005.2.20)  21 FEB 2005 14:49
entering extended mode
**latex1.tex
(./latex1.tex
LaTeX2e <2003/12/01>
Babel <v3.8d> and hyphenation patterns for american, french, german, ngerman, b
ahasa, basque, bulgarian, catalan, croatian, czech, danish, dutch, esperanto, e
stonian, finnish, greek, icelandic, irish, italian, latin, magyar, norsk, polis
h, portuges, romanian, russian, serbian, slovak, slovene, spanish, swedish, tur
kish, ukrainian, nohyphenation, loaded.
(/usr/local/stow/tetex-3.0/share/texmf-dist/tex/latex/base/article.cls
Document Class: article 2004/02/16 v1.4f Standard LaTeX document class
(/usr/local/stow/tetex-3.0/share/texmf-dist/tex/latex/base/size10.clo
File: size10.clo 2004/02/16 v1.4f Standard LaTeX file (size option)
)
\c@part=\count79
\c@section=\count80
\c@subsection=\count81
\c@subsubsection=\count82
\c@paragraph=\count83
\c@subparagraph=\count84
\c@figure=\count85
\c@table=\count86
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)
(./latex1.aux)
\openout1 = `latex1.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 2.
LaTeX Font Info:    ... okay on input line 2.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 2.
LaTeX Font Info:    ... okay on input line 2.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 2.
LaTeX Font Info:    ... okay on input line 2.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 2.
LaTeX Font Info:    ... okay on input line 2.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 2.
LaTeX Font Info:    ... okay on input line 2.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 2.
LaTeX Font Info:    ... okay on input line 2.
\QWSQXvl=\skip43
LaTeX Font Info:    External font `cmex10' loaded for size
(Font)              <7> on input line 3.
LaTeX Font Info:    External font `cmex10' loaded for size
(Font)              <5> on input line 3.
\QWSQXel=\skip44
\JPBJBvl=\skip45
\JPBJBel=\skip46
\UJTTBa=\skip47
\UJTTBb=\skip48

Overfull \hbox (24.79248pt too wide) in paragraph at lines 12--33
[][] 
 []

[1

] (./latex1.aux) ) 
Here is how much of TeX's memory you used:
 212 strings out of 94500
 2317 string characters out of 1174952
 50348 words of memory out of 1000000
 3466 multiletter control sequences out of 10000+50000
 3640 words of font info for 14 fonts, out of 500000 for 2000
 580 hyphenation exceptions out of 1000
 23i,8n,17p,130b,107s stack positions out of 1500i,500n,5000p,200000b,5000s

Output written on latex1.dvi (1 page, 1724 bytes).
