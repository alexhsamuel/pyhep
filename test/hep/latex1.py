#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.latex import *
from   hep.num import Statistic
import os
import sys

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

table = Table(
    TextColumn("Label", "center"),
    StatisticColumn("Stat 1", 0, 2, "%"),
    StatisticColumn("Stat 2", 1, 2, 1),
    PatternColumn("Pattern 1", "B |<xxxxxx| M |>xxxxxx| E"),
    )

table.append("first row",  "1.0000+-.120", " 100+-2.1", ("apple", "water"))
table.append("second row", "0.5000+-.012", "  10+-2.2", ("pear", "milk"))
table.append("thrid row",  "0.1000+-.001", "   1+-2.3", ("orange", "beer"))
table.append("fourth row", "0.0100+-.100", "  -1+-2.4", ("banana", "juice"))
table.append("fifth row",  "0.0010+-.020", " -10+-2.5", ("grape", "wine"))
table.append("sixth row",  "0.0061+-.024", "-100+-2.6", ("quince", "coffee"))
table.append("sevent row", "0.0042+-.026", "   0+-2.7", ("plum", "soda"))
table.append("eight row",  "0.0050+-.032", "   0+-2.8", ("apricot", "tea"))
             
output = file("latex1.tex", "w")
print >> output, r"\documentclass{article}"
print >> output, r"\begin{document}"
print >> output, table.generate()
print >> output, r"\end{document}"
del output

os.system("latex --interaction nonstopmode latex1.tex")

