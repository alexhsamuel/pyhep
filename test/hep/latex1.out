Traceback (most recent call last):
  File "latex1.py", line 33, in ?
    print >> output, table.generate()
  File "/home/samuel/dev/pyhep/py/hep/latex.py", line 531, in generate
    result = "".join([ c.generateSetup(self) for c in self.columns ])
  File "/home/samuel/dev/pyhep/py/hep/latex.py", line 205, in generateSetup
    lengths.append[length]
TypeError: unsubscriptable object
